glogger/handler.py

```diff
@@ -0,0 +1,277 @@
+        :param flush_interval: Seconds between sending batches.
+        :param flush_threshold: Force send when buffer utilization reaches this percentage.
+        :param overflow_drop_factor: Percentage of messages to be dropped when buffer becomes full.
+        :param max_send_tries: Number of times to retry sending a batch if sending fails.
+        """
+        super().__init__(logging.DEBUG)
+        self.max_message_size = max_message_size  # maximum message size
+        self.flush_interval = flush_interval  # maximum amount of seconds between flushes
+        self.flush_threshold = flush_threshold  # force flush if buffer size reaches this percentage of capacity
+        self.max_send_tries = max_send_tries  # maximum number of times to retry sending logs if request fails
+        self.messages_buffer = MessagesBuffer(max_total_length, overflow_drop_factor)
+        self.messages_buffer.head_serial_no = continue_from
+        self.stop_event = threading.Event()
+        self.time_fn = time.time
+        self.jsonify = JSONEncoder(separators=(",", ":")).encode  # compact, no whitespace
+        self.uri = f"{self.scheme}://{server_address}/api/v1/logs"
+        self.session = requests.Session()
+        self.session.headers.update(
+            {
+                "X-Application-Name": application_name,
+                "X-Token": auth_token,
+            }
+        )
+        self.last_flush_time = 0.0
+        self.flush_thread = threading.Thread(target=self._flush_loop, daemon=True, name="Logs flusher")
+        self.start()
+
+    def start(self) -> None:
+        """Called to start the flush thread."""
+        assert not self.stop_event.is_set(), "restart not allowed!"
+        self.flush_thread.start()
+
+    def emit(self, record: LogRecord) -> None:
+        # Called while lock is acquired
+        prev_dropped = self.messages_buffer.dropped
+        self.messages_buffer.append(self.format(record))
+        new_dropped = self.messages_buffer.dropped - prev_dropped
+        if new_dropped:
+            self.warn_dropped(new_dropped)
+
+    def format(self, record: LogRecord) -> str:
+        super().format(record)
+        formatted_timestamp = datetime.utcfromtimestamp(record.created).isoformat()
+        d = {
+            **self.get_extra_fields(record),
+            "message": record.message,
+            "serial_no": self.messages_buffer.next_serial_no,
+            "severity": record.levelno,
+            "timestamp": formatted_timestamp,
+            "logger_name": record.name,
+        }
+        s = self.jsonify(d)
+        if len(s) > self.max_message_size:
+            self.truncate(d)
+            s = self.jsonify(d)
+            assert len(s) <= self.max_message_size, "did not truncate enough!"
+        return s
+
+    def get_extra_fields(self, record: LogRecord) -> dict:
+        """Override to add extra fields to formatted record."""
+        return {}
+
+    def truncate(self, d):
+        minimum = 80
```

Comment:
I think the logic is a bit complex - would change to:
1. jsonify the string and understand if over the limit.
2. If over the limit understand by how much, then truncate from only the message field - that's the only thing that should do it big enough
3. Then assert that it works or drop the message (and indicate that it was dropped)

Reply:
The message field is actually not likely to be very large since almost all our message our short static messages. The long parts are stacktraces usually

Reply:
`exception` and `stack_trace`